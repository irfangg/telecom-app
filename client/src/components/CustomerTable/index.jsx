import { useCallback, useState } from "react";
import Table from "react-bootstrap/Table";
import { Button } from "react-bootstrap";
import ChoosePlan from "../Plan/ChoosePlan";
import ModifyPlan from "../Plan/ModifyPlan";
import RenewPlan from "../Plan/RenewPlan";

function CustomerTable({ tableData, setTableData }) {
  const [selectedRow, setSelectedRow] = useState(null);
  const [showChoosePlan, setShowChoosePlan] = useState(false);
  const [showRenewPlan, setShowRenewPlan] = useState(false);
  const [showModifyPlan, setShowModifyPlan] = useState(false);

  const handleChoosePlan = useCallback((row) => {
    setShowChoosePlan(true);
    setSelectedRow(row);
  }, []);
  const handleRenewPlan = useCallback((row) => {
    setShowRenewPlan(true);
    setSelectedRow(row);
  }, []);
  const handleModifyPlan = useCallback((row) => {
    setShowModifyPlan(true);
    setSelectedRow(row);
  }, []);

  return (
    <>
      <Table striped bordered hover>
        <thead>
          <tr>
            <th>Id</th>
            <th>Customer Name</th>
            <th>Email</th>
            <th>Registration Date</th>
            <th>Assigned number</th>
            <th>Plan Details</th>
          </tr>
        </thead>
        <tbody>
          {tableData?.map((data, i) => {
            return (
              <tr key={data?.id || i}>
                <td>{data?.id}</td>
                <td>{data?.name}</td>
                <td>{data.email}</td>
                <td>{data.registrationDate}</td>
                <td>{data.assignedMobileNumber}</td>
                <td>
                  {data.plan ? (
                    <Button
                      variant="primary"
                      size="sm"
                      onClick={() =>
                        data.plan?.planStatus
                          ? handleModifyPlan(data)
                          : handleRenewPlan(data)
                      }
                    >
                      {data.plan?.planStatus ? "Modify plan" : "Renew plan"}
                    </Button>
                  ) : (
                    <Button
                      variant="primary"
                      size="sm"
                      onClick={() => handleChoosePlan(data)}
                    >
                      Choose plan
                    </Button>
                  )}
                </td>
              </tr>
            );
          })}
        </tbody>
      </Table>
      <ChoosePlan
        {...{
          showChoosePlan,
          setShowChoosePlan,
          customer: selectedRow,
          setTableData,
        }}
      />
      <ModifyPlan
        {...{
          showModifyPlan,
          setShowModifyPlan,
          customer: selectedRow,
          setTableData,
        }}
      />
      <RenewPlan
        {...{
          showRenewPlan,
          setShowRenewPlan,
          customer: selectedRow,
          setTableData,
        }}
      />
    </>
  );
}

export default CustomerTable;
