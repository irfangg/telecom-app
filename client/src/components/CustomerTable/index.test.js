import renderer from 'react-test-renderer';
import CustomerTable from '.';
import AuthProvider from '../../context/AuthProvider';

describe('Test Dashboard', ()=>{
    it('If new user generate snap shot to choose plan', () => {
        const tableData = [{
            "name": "irfangg",
            "dob": "2024-05-12",
            "email": "irfan@gmail.com",
            "adharNumber": "213",
            "registrationDate": 1732110946770,
            "assignedMobileNumber": "122",
            "plan": null,
            "id": "cb3e"
        }];
        const setTableData = jest.fn();
        const component = renderer.create(
            <AuthProvider>
                <CustomerTable tableData={tableData} setTableData={setTableData} />
            </AuthProvider>
        );
        expect(component).toMatchSnapshot();
    });
    it('If existing user and plan status is active generate snapshot for modify plan', () => {
        const tableData = [{
            "name": "irfangg",
            "dob": "2024-05-12",
            "email": "irfan@gmail.com",
            "adharNumber": "213",
            "registrationDate": 1732110946770,
            "assignedMobileNumber": "122",
            "plan": {
                "planName": "Gold",
                "planCost": 299,
                "validity": 180,
                "planStatus": true
            },
            "id": "cb3e"
        }];
        const setTableData = jest.fn();
        const component = renderer.create(
            <AuthProvider>
                <CustomerTable tableData={tableData} setTableData={setTableData} />
            </AuthProvider>
        );
        expect(component).toMatchSnapshot();
    });
    it('If existing user and plan status is not active generate snapshot for renew plan', () => {
        const tableData = [{
            "name": "irfangg",
            "dob": "2024-05-12",
            "email": "irfan@gmail.com",
            "adharNumber": "213",
            "registrationDate": 1732110946770,
            "assignedMobileNumber": "122",
            "plan": {
                "planName": "Gold",
                "planCost": 299,
                "validity": 180,
                "planStatus": false
            },
            "id": "cb3e"
        }];
        const setTableData = jest.fn();
        const component = renderer.create(
            <AuthProvider>
                <CustomerTable tableData={tableData} setTableData={setTableData} />
            </AuthProvider>
        );
        expect(component).toMatchSnapshot();
    })
});