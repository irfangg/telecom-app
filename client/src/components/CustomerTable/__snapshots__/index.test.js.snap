// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Dashboard If existing user and plan status is active generate snapshot for modify plan 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        Id
      </th>
      <th>
        Customer Name
      </th>
      <th>
        Email
      </th>
      <th>
        Registration Date
      </th>
      <th>
        Assigned number
      </th>
      <th>
        Plan Details
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        cb3e
      </td>
      <td>
        irfangg
      </td>
      <td>
        irfan@gmail.com
      </td>
      <td>
        1732110946770
      </td>
      <td>
        122
      </td>
      <td>
        <button
          className="btn btn-primary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Modify plan
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Test Dashboard If existing user and plan status is not active generate snapshot for renew plan 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        Id
      </th>
      <th>
        Customer Name
      </th>
      <th>
        Email
      </th>
      <th>
        Registration Date
      </th>
      <th>
        Assigned number
      </th>
      <th>
        Plan Details
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        cb3e
      </td>
      <td>
        irfangg
      </td>
      <td>
        irfan@gmail.com
      </td>
      <td>
        1732110946770
      </td>
      <td>
        122
      </td>
      <td>
        <button
          className="btn btn-primary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Renew plan
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Test Dashboard If new user generate snap shot to choose plan 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        Id
      </th>
      <th>
        Customer Name
      </th>
      <th>
        Email
      </th>
      <th>
        Registration Date
      </th>
      <th>
        Assigned number
      </th>
      <th>
        Plan Details
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        cb3e
      </td>
      <td>
        irfangg
      </td>
      <td>
        irfan@gmail.com
      </td>
      <td>
        1732110946770
      </td>
      <td>
        122
      </td>
      <td>
        <button
          className="btn btn-primary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Choose plan
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
